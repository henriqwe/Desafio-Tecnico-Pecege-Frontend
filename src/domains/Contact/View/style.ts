import styled from 'styled-components'
import * as Icons from '@heroicons/react/24/outline'

export const Container = styled.section`
  position: relative;
  max-height: 100vh;
  min-height: 100vh;
  display: flex;
  flex: 1;
  flex-direction: column;
  overflow-x: none;
`

export const ChevronLeftIcon = styled(Icons.ChevronLeftIcon)`
  width: 2rem;
  height: 2rem;
`

export const EllipsisVerticalIcon = styled(Icons.EllipsisVerticalIcon)`
  width: 2rem;
  height: 2rem;
`

export const ContactHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`
export const AvatarWrapper = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
  padding-top: 1rem;
  gap: 0.4rem;
`
export const AvatarName = styled.p`
  font-size: 1.1rem;
  color: rgb(15 23 42);
  font-weight: 800;
`
export const AvatarPhone = styled.p`
  font-size: 0.9rem;
  color: rgb(100 116 139);
  font-weight: 400;
`
export const ContactHeaderWrapper = styled.div`
  background-color: rgba(100, 116, 139, 0.1);
  padding: 0.8rem;
  padding-top: 1rem;
`
export const UserIcon = styled(Icons.UserIcon)`
  width: 1.5rem;
  height: 1.5rem;
`
export const MapPinIcon = styled(Icons.MapPinIcon)`
  width: 1.5rem;
  height: 1.5rem;
`
export const BuildingStorefrontIcon = styled(Icons.BuildingStorefrontIcon)`
  width: 1.5rem;
  height: 1.5rem;
`
export const InfoTitle = styled.p`
  font-size: 0.9rem;
  color: rgb(15 23 42);
  font-weight: 800;
`
export const InfoDescription = styled.p`
  font-size: 1.1rem;
  color: rgb(100 116 139);
  font-weight: 400;
  padding-left: 1rem;
`
export const TabItemContainer = styled.section`
  display: flex;
  flex: 1;
  flex-direction: column;
  margin: 0.8rem;
  box-shadow: 0 2px 5px rgba(0, 0, 0, 0.25);
`
export const InfoWrapper = styled.div<{ variant: 'white' | 'gray' }>`
  display: flex;
  flex: 1;
  gap: 0.2rem;
  padding: 0.7rem;
  flex-direction: column;
  background-color: ${({ variant }) =>
    variant === 'white' ? '#fff' : '#F1F5F9'};
`
export const DropdownOptionContent = styled.div`
  display: flex;
  gap: 0.5rem;
  align-items: end;
  justify-content: space-between;
  width: 100%;
`
export const PencilIcon = styled(Icons.PencilIcon)`
  width: 1.3rem;
  height: 1.3rem;
`
export const TrashIcon = styled(Icons.TrashIcon)`
  width: 1.3rem;
  height: 1.3rem;
  color: red;
`
export const MapContainer = styled.div`
  width: 100%;
  height: 20rem;
`
