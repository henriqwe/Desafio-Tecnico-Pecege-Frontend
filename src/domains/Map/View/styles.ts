import styled from 'styled-components'

export const MapViewContainer = styled.div`
  width: 100vw;
  height: 100vh;
`
