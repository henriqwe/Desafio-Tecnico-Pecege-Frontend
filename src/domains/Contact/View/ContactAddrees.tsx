import * as S from './style'
import type { IContact } from '&types/contact'
import { Map } from '&components/Map/Map'
import { motion } from 'framer-motion'

export function ContactAddrees({ contact }: { contact: IContact }) {
  return (
    <S.TabItemContainer data-testid={'tab-container'}>
      <motion.div
        initial={{ x: -300, opacity: 0 }}
        animate={{ x: 0, opacity: 1 }}
        exit={{ x: -300, opacity: 0 }}
        transition={{
          type: 'spring',
          stiffness: 260,
          damping: 40
        }}
        style={{
          width: '100%',
          display: 'flex'
        }}
      >
        <S.InfonContainer>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>Rua</S.InfoTitle>
            <S.InfoDescription data-testid={'info-street'}>
              {contact.address.street || '-'}
            </S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="gray">
            <S.InfoTitle>NÃºmero</S.InfoTitle>
            <S.InfoDescription data-testid={'info-suite'}>
              {contact.address.suite || '-'}
            </S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>Cidade</S.InfoTitle>
            <S.InfoDescription data-testid={'info-city'}>
              {contact.address.city || '-'}
            </S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="gray">
            <S.InfoTitle>CEP</S.InfoTitle>
            <S.InfoDescription data-testid={'info-zipcode'}>
              {contact.address.zipcode || '-'}
            </S.InfoDescription>
          </S.InfoWrapper>
        </S.InfonContainer>
      </motion.div>

      {contact.address.geo.lat && contact.address.geo.lng && (
        <motion.div
          initial={{ x: 300, opacity: 0 }}
          animate={{ x: 0, opacity: 1 }}
          exit={{ x: 300, opacity: 0 }}
          transition={{
            type: 'spring',
            stiffness: 260,
            damping: 40
          }}
          style={{
            width: '100%',
            height: '100%',
            display: 'flex',
            backgroundColor: 'red'
          }}
        >
          <S.MapContainer>
            <Map contacts={[contact]} showInfoWindow={false} />
          </S.MapContainer>
        </motion.div>
      )}
    </S.TabItemContainer>
  )
}
