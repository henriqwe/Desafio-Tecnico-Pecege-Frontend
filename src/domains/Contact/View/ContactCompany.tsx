import * as S from './style'
import type { IContact } from '&types/contact'
import { motion } from 'framer-motion'

export function ContactCompany({ contact }: { contact: IContact }) {
  return (
    <motion.div
      initial={{ y: 300, opacity: 0 }}
      animate={{ y: 0, opacity: 1 }}
      exit={{ y: 300, opacity: 0 }}
      transition={{
        type: 'spring',
        stiffness: 260,
        damping: 40
      }}
      style={{ width: '100%', height: '100%' }}
    >
      <S.TabItemContainer>
        <S.InfonContainer>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>Nome</S.InfoTitle>
            <S.InfoDescription>{contact.company.name || '-'}</S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="gray">
            <S.InfoTitle>Frase de efeito</S.InfoTitle>
            <S.InfoDescription>
              {contact.company.catchPhrase || '-'}
            </S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>Estratégia de negócio</S.InfoTitle>
            <S.InfoDescription>{contact.company.bs || '-'}</S.InfoDescription>
          </S.InfoWrapper>
        </S.InfonContainer>
      </S.TabItemContainer>
    </motion.div>
  )
}
