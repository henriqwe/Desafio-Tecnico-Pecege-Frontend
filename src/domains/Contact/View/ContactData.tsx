import * as S from './style'
import type { IContact } from '&types/contact'
import { motion } from 'framer-motion'

export function ContactData({ contact }: { contact: IContact }) {
  return (
    <motion.div
      initial={{ y: 300, opacity: 0 }}
      animate={{ y: 0, opacity: 1 }}
      exit={{ y: 300, opacity: 0 }}
      transition={{
        type: 'spring',
        stiffness: 260,
        damping: 40
      }}
      style={{ width: '100%', height: '100%' }}
    >
      <S.TabItemContainer>
        <S.InfonContainer>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>Nome</S.InfoTitle>
            <S.InfoDescription>{contact.name || '-'}</S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="gray">
            <S.InfoTitle>Telefone</S.InfoTitle>
            <S.InfoDescription>{contact.phone || '-'}</S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="white">
            <S.InfoTitle>E-mail</S.InfoTitle>
            <S.InfoDescription>{contact.email || '-'}</S.InfoDescription>
          </S.InfoWrapper>
          <S.InfoWrapper variant="gray">
            <S.InfoTitle>Website</S.InfoTitle>
            <S.InfoDescription>{contact.website || '-'}</S.InfoDescription>
          </S.InfoWrapper>
        </S.InfonContainer>
      </S.TabItemContainer>
    </motion.div>
  )
}
