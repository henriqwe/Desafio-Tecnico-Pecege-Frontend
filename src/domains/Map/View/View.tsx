import { FloatingActionButton } from '&components/FloatingActionButton/FloatingActionButton'
import { Map } from '&components/Map/Map'
import type { IContact } from '&types/contact'
import { fetchContacts } from '&services/queries/fetchContacts'
import { useQuery, useQueryClient } from '@tanstack/react-query'
import * as S from './styles'

export function MapView() {
  const queryClient = useQueryClient()

  async function handleFetch() {
    const contacts = queryClient.getQueryData(['contacts']) as IContact[]
    if (contacts) {
      return contacts
    }
    return fetchContacts()
  }
  const contactsQuery = useQuery({
    queryKey: ['contacts'],
    queryFn: () => handleFetch(),
    refetchOnWindowFocus: false,
    retry: false
  })
  return (
    <S.MapViewContainer data-testid={'mapview-container'}>
      {!contactsQuery.data ? (
        <div></div>
      ) : (
        // <Loading />
        <>
          <Map contacts={contactsQuery.data} />
          <FloatingActionButton />
        </>
      )}
    </S.MapViewContainer>
  )
}
