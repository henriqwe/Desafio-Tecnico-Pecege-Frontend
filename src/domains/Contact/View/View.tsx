import { useNavigate, useParams } from 'react-router-dom'
import { Loading } from '&components/Loading/Loading'
import { routes } from '&utils/routes'
import { Avatar } from '&components/Avatar/Avatar'
import { useQueryClient, useMutation } from '@tanstack/react-query'
import { notification } from '&utils/notification'
import * as S from './style'
import { Tab, TabItem } from '&components/Tab/Tabs'
import { useState } from 'react'
import { ContactData } from './ContactData'
import { ContactAddrees } from './ContactAddrees'
import { ContactCompany } from './ContactCompany'
import { Dropdown } from '&components/Dropdown/Dropdown'
import { Modal } from '&components/Modal/Modal'
import { fakePromise } from '&utils/fakePromise'
import { useContact } from '&hooks/useContact'

type TSelectedTab = 0 | 1 | 2

export function ViewContact() {
  const { id } = useParams()

  const contact = useContact(id as string)

  const navigate = useNavigate()

  const onTabSelected = (index: number) => {
    setSelectedTab(index as TSelectedTab)
  }

  const [selectedTab, setSelectedTab] = useState<TSelectedTab>(0)
  const [openModal, setOpenModal] = useState(false)
  const queryClient = useQueryClient()

  const removeContact = useMutation({
    mutationFn: (id: number) => fakePromise(),
    onSuccess: (result, id) => {
      navigate(routes.home.path)
      queryClient.setQueryData(['contacts'], (old: any) =>
        old.filter((contact: any) => contact.id !== id)
      )

      notification('Contato removido com sucesso', 'success')
    },
    onError: (error: Error) => notification(error.message, 'error')
  })
  if (contact.error) {
    navigate(routes.home.path)
    notification(contact.error.message, 'error')
    return <></>
  }

  const TabsContent = {
    0: <ContactData contact={contact.data!} />,
    1: <ContactAddrees contact={contact.data!} />,
    2: <ContactCompany contact={contact.data!} />
  }
  const dropdownOption = [
    {
      content: (
        <S.DropdownOptionContent>
          Editar
          <S.PencilIcon />
        </S.DropdownOptionContent>
      ),
      fn: () => navigate(routes.editContact.path(contact.data?.id!))
    },
    {
      content: (
        <S.DropdownOptionContent>
          Remover
          <S.TrashIcon />
        </S.DropdownOptionContent>
      ),
      fn: () => setOpenModal(true)
    }
  ]
  return (
    <S.Container>
      {contact.isLoading || !contact.data ? (
        <S.LoadingContainer>
          <Loading />
        </S.LoadingContainer>
      ) : (
        <>
          <S.ContactHeaderWrapper>
            <S.ContactHeader>
              <div onClick={() => navigate(routes.home.path)}>
                <S.ChevronLeftIcon />
              </div>
              <h1>Contato</h1>

              <Dropdown options={dropdownOption} />
            </S.ContactHeader>

            <Modal
              action={() => removeContact.mutate(contact.data?.id!)}
              title="Deseja realmente remover esse contato?"
              actionsText="Sim"
              cancelText="Não"
              open={openModal}
              onOpenChange={setOpenModal}
            />
            <S.AvatarWrapper>
              <Avatar name={contact.data.name} variant="md" />
              <S.AvatarName>{contact.data.name}</S.AvatarName>
              <S.AvatarPhone>{contact.data.phone}</S.AvatarPhone>
            </S.AvatarWrapper>
          </S.ContactHeaderWrapper>
          <div>
            <Tab onTabSelected={onTabSelected}>
              <TabItem>
                <S.UserIcon />
                Dados
              </TabItem>
              <TabItem>
                <S.MapPinIcon />
                Endereço
              </TabItem>
              <TabItem>
                <S.BuildingStorefrontIcon />
                Empresa
              </TabItem>
            </Tab>
            <div>{TabsContent[selectedTab]}</div>
          </div>
        </>
      )}
    </S.Container>
  )
}
