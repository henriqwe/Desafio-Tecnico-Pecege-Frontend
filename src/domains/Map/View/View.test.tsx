import { render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { MapView } from './View'
import { initialize } from '@googlemaps/jest-mocks'
import { TestProviders } from '&utils/tests/TestProviders'

jest.mock('@tanstack/react-query', () => ({
  ...jest.requireActual('@tanstack/react-query'),
  useQueryClient: () => ({
    setQueryData: jest.fn(),
    getQueryData: jest.fn()
  })
}))

beforeEach(() => {
  initialize()
})

describe('MapView', () => {
  it('should render the contact list correctly', async () => {
    const { container, findByTestId } = render(
      <TestProviders>
        <MapView />
      </TestProviders>
    )

    await waitFor(async () => findByTestId('mapview-container'))
    const mapContainer = await findByTestId('mapview-container')
    expect(container).toMatchSnapshot()
    expect(mapContainer).toBeTruthy()
  })
})
