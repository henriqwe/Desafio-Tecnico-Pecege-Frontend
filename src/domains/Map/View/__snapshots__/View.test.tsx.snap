// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView should render the contact list correctly 1`] = `
<div>
  <div
    class="sc-jrkPcm gOcgPd"
    data-testid="mapview-container"
  >
    <div />
  </div>
</div>
`;
